### PR TITLE
Add documentation for memory management

### PR DESCRIPTION
**Pull request by xiaoxmeng**

### CODE REVIEW

File: velox/docs/develop/memory.rst [hunk 1]
```diff
@@ -0,0 +1,890 @@
...
+operator pool can proceed with the actual memory allocation from memory
+allocator. However, if the memory arbitration fails, then the query which has
+the largest capacity in the system is chosen to fail with query memory capacity
+exceeded error (local OOM). The failed query may or may not be the requestor
+query pool itself.
+
+The memory allocator does the actual memory allocation from its own managed
+memory space in units of machine pages (4KB). It tracks the amount of allocated
+memory, and returns an error if an allocation request exceeds the system memory
+limit. This enables explicit control on RSS of Velox memory usage to help
```
Comment (duanmeng) [inline velox/docs/develop/memory.rst:91]: s/RSS/RSS (Resident Set Size) ? I used to think RSS was a Remote Shuffle Service :)

File: velox/docs/develop/memory.rst [hunk 2]
```diff
@@ -0,0 +1,890 @@
...
+
+The implementation of MemoryPool::incrementReservationThreadSafe:
+
+#. A non-root memory pool calls its parent pool’s *incrementReservationThreadSafe*
+   method recursively to propagate the reservation request all the way up to
+   the root memory pool
+
+#. Check *MemoryPool::incrementReservationThreadSafe* result from the parent pool:
+
+   #. If the function returns true, the reservation succeeds from the root
```
Comment (duanmeng) [inline velox/docs/develop/memory.rst:352]: s/#/a then rest sequence id would be b, c, ...

File: velox/docs/develop/memory.rst [hunk 3]
```diff
@@ -0,0 +1,890 @@
...
+      pool. We need to retry from the leaf memory pool again by returning false
+      to *MemoryPoolImpl::reserveThreadSafe*
+   #. If the memory reservation fails at the root memory pool, the function
+      expects a query memory capacity exceeded exception thrown and the memory
+      allocation fails
+
+#. Call *MemoryPool::maybeIncrementReservationLocked* to try to increment the
+   reservation and check the result:
+
+   #. For a non-root memory pool, this should always succeed as we only check
```
Comment (duanmeng) [inline velox/docs/develop/memory.rst:365]: ditto

File: velox/docs/develop/memory.rst [hunk 4]
```diff
@@ -0,0 +1,890 @@
...
+   of the root query pools as the memory arbitration candidates. The memory
+   manager keeps the candidate query pools alive during the memory arbitration
+   process
+#. The memory arbitrator serializes the memory arbitration processing with one
+   request at a time to ensure a consistent view of memory capacity allocated
+   among queries. The memory arbitrator might receive concurrent arbitration
+   requests from different queries or even from different driver threads of the
+   same query. For each memory arbitration request:
+
+   #. The memory arbitrator invokes *MemoryPool::enterArbitration( method of the
```
Comment (duanmeng) [inline velox/docs/develop/memory.rst:574]: ditto

File: velox/docs/develop/memory.rst (hunk 1 above)
Comment (xiaoxmeng) [reply velox/docs/develop/memory.rst:91]: Resident Set Size is abbreviated as RSS at the first place in this doc? So we assume RSS is Resident Set Size? Thanks!

**Review by duanmeng — APPROVED** [review]
Nice Document 👍

**Comment by mbasmanova** [comment]
CC: @aditi-pandit @majetideepak @rui-mo

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Meng, thank you for putting this documentation together. Please, make sure it renders well before landing.

File: velox/docs/develop/memory.rst [hunk 5]
```diff
@@ -0,0 +1,890 @@
...
+.. image:: images/memory-manager.png
+    :width: 600
+    :align: center
+    :alt: Memory Manager
+
+The memory manager is created on server startup with provided
+*MemoryManagerOption*. It creates a memory allocator instance to manage the
+physical memory allocations for both query memory allocated through memory pool
+and cache memory allocated through the file cache. It ensures the total
+allocated memory is within the system memory limit (specified by
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:128]: a leading * is missing in "MemoryManagerOptions::allocatorCapacity*"

File: velox/docs/develop/memory.rst [hunk 6]
```diff
@@ -0,0 +1,890 @@
...
+MemoryManagerOptions::allocatorCapacity*). The memory manager also creates a
+memory arbitrator instance to arbitrate the memory capacity among running
+queries. It ensures the total allocated query memory capacity is within the
+query memory limit (specified by *MemoryManagerOptions::arbitratorCapacity*). The
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:132]: Maybe include a digram showing relationship between different limits: allocatorCapacity vs. arbitratorCapacity vs. total memory available for the whole process.

File: velox/docs/develop/memory.rst [hunk 7]
```diff
@@ -0,0 +1,890 @@
...
+limit no matter if it is for system operation or for user query execution. In
+practice, we shall reserve some space from the memory allocator to compensate
+for such system memory usage. We can do that by configuring a smaller query
+memory limit (*MemoryManagerOptions::arbitratorCapacity*) than system memory
+limit (*MemoryManagerOptions::allocatorCapacity*) (refer to `OOM prevention section <#server-oom-prevention>`_
+for detail).
+
+Memory System Setup
+^^^^^^^^^^^^^^^^^^^
+Here is the code block from Prestissimo that initializes the Velox memory system:
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:180]: nit: add empty line

File: velox/docs/develop/memory.rst [hunk 8]
```diff
@@ -0,0 +1,890 @@
...
+* L11: set the memory arbitrator capacity (query memory limit) from
+  Prestissimo system config
+* L14: creates the process-wide memory manager which creates memory
+  allocator and arbitrator inside based on MemoryManagerOptions initialized from previous steps
+* L15-19: creates the file cache if it is enabled in Prestissimo system
+  config
+
+Memory Pool
+-----------
+The memory pool provides memory allocation functions for query execution. It
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:225]: nit: add empty line

File: velox/docs/develop/memory.rst [hunk 9]
```diff
@@ -0,0 +1,890 @@
...
+
+.. image:: images/memory-pool.png
+    :width: 500
+    :align: center
+    :alt: Memory Pool
+
+Query allocates memory from the operator pool at the leaf of the tree and
+propagates the memory usage all the way up to the query pool at the root of the
+tree to check if the memory usage has exceeded the per-query memory limit or
+not. The memory allocation always happens at the leaf operator pool, the
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:250]: Might be helpful to discuss whether leaf memory pools are thread-safe or not.

File: velox/docs/develop/memory.rst [hunk 10]
```diff
@@ -0,0 +1,890 @@
...
+simplify the memory pool management: one is *LEAF* type which only allows the
+memory allocations and the other is *AGGREGATE* type which aggregates the
+memory usage from all its children but is not allowed to allocate memory
+directly. Hence, the operator pool is *LEAF* type and all the others are
+*AGGREGATE* type. We only enforce the memory limit check at the root query
+pool.
+
+Memory Usage Tracking
+^^^^^^^^^^^^^^^^^^^^^
+To track query memory usage, a leaf operator pool needs to propagate the memory
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:263]: Nice explanation.

File: velox/docs/develop/memory.rst [hunk 11]
```diff
@@ -0,0 +1,890 @@
...
+queries to achieve fair memory sharing and prevent a query from running out of
+its memory limit. To arbitrate memory capacity between running queries, the
+memory arbitrator needs to be able to reclaim the used memory from a query
+through techniques such as `disk spilling <https://facebookincubator.github.io/velox/develop/spilling.html>`_, and then transfer the freed memory
+between queries by adjusting their memory pool’s capacities accordingly (see
+`memory arbitration process section <#memory-arbitration-process>`_ for detail).
+
+The *MemoryArbitrator* is defined to support different implementations for
+different query systems. As for now, we implement *SharedArbitrator* for both
+Prestissimo and Prestission-on-Spark. `Gulten <https://github.com/oap-project/gluten>`_ implements its own memory
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:534]: typo: Prestission-on-Spark -> Presto-on-Spark or Prestissimo-on-Spark

**Comment by rui-mo** [comment]
cc @zhztheplayer

File: velox/docs/develop/memory.rst [hunk 12]
```diff
@@ -0,0 +1,890 @@
...
+actual data processing to allow the memory arbitrator to reclaim memory.
+
+Memory Allocator
+----------------
+The memory allocator manages the physical memory allocations for both query
+memory allocated through memory pool and cache memory allocated directly from
+file cache. The memory allocator ensures the total allocated memory is always
+within the system memory limit. *MemoryAllocator* defines the memory allocator
+interface. We have two allocator implementations: *MallocAllocator* delegates
+all the memory allocations to std::malloc which is simple and reliable. We
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:687]: nit: might be helpful to clarify that it is possible to configure the application to use jemalloc as well.

File: velox/docs/develop/memory.rst [hunk 13]
```diff
@@ -0,0 +1,890 @@
...
+provide it as the default option but we believe it has the issue with RSS
+variation caused by memory fragmentation. Therefore we built *MMapAllocator* to
+manage the physical memory allocations using the std::mmap to have explicit
+control on RSS of the Velox memory usage. We haven't yet confirmed whether
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:691]: drop "of the Velox memory usage"

File: velox/docs/develop/memory.rst [hunk 14]
```diff
@@ -0,0 +1,890 @@
...
+&MmapAllocator& works better than &MallocAllocator*, but we are able to run a
```
Comment (mbasmanova) [inline velox/docs/develop/memory.rst:692]: & and * are mixed

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Nice document. Thank you for taking the time to put it together and iterate. I have a few nits to consider for a follow-up.

File: velox/docs/develop/memory.rst (hunk 9 above)
Comment (xiaoxmeng) [reply velox/docs/develop/memory.rst:250]: It is mentioned in memory usage tracking.

File: velox/docs/develop/memory.rst (hunk 12 above)
Comment (xiaoxmeng) [reply velox/docs/develop/memory.rst:687]: Added "Users can choose the
allocator for their application by setting *MemoryManagerOptions::useMmapAllocator*
(see for `memory system setup section <#memory-system-setup>`_ for example)" at end of this paragraph

File: velox/docs/develop/memory.rst (hunk 6 above)
Comment (xiaoxmeng) [reply velox/docs/develop/memory.rst:132]: Good point, I will add one in followup.

**Pull request closed without merging**